superduperdb/components/graph.py

```diff
@@ -0,0 +1,255 @@
+        self.identifier = name
+        self.inputs = []
+
+    def predict(self, x):
+        return x
+
+
+@dc.dataclass(kw_only=True)
+class Graph(Component, _Predictor):
+    models: t.List[Model] = dc.field(default_factory=list)
+    edges: t.List[t.Tuple[str, str, t.Tuple[int, str]]] = dc.field(default_factory=list)
+    input: Model
+    outputs: t.Optional[t.List[str]] = dc.field(default_factory=list)
+    _DEFAULT_ARG_WEIGHT: t.ClassVar[tuple] = (-1, '_base')
+    type_id: t.ClassVar[str] = 'graph'
+
+    def __post_init__(self, artifacts):
+        self.G = nx.DiGraph()
+        self.nodes = {}
+        self._node_output_cache = {}
+        self.version = 0
+        self._db = None
+
+        self.connect_root(self.input)
+
+        # Load the models and edges into a di graph
+        models = {m.identifier: m for m in self.models}
+        if self.edges and models:
+            for connection in self.edges:
+                u, v, on = connection
+                self.connect(
+                    models[u],
+                    models[v],
+                    on=on,
+                    update_edge=False,
+                )
+        super().__post_init__(artifacts=artifacts)
+
+    def connect_root(self, v: Model):
+        if v.identifier not in self.nodes:
+            self.nodes[v.identifier] = v
+
+        self.nodes[self.identifier] = GraphModel(self.identifier)
+        self.G.add_edge(self.identifier, v.identifier, weight=self._DEFAULT_ARG_WEIGHT)
+
+    def connect(
+        self,
+        u: Model,
+        v: Model,
+        on: t.Optional[t.Tuple[int, str]] = None,
+        update_edge: t.Optional[bool] = True,
+    ):
+        assert isinstance(u, Model)
+        assert isinstance(v, Model)
+
+        if u.identifier not in self.nodes:
+            self.nodes[u.identifier] = u
+            self.G.add_node(u.identifier)
+
+        if v.identifier not in self.nodes:
+            self.nodes[v.identifier] = v
+            self.G.add_node(v.identifier)
+
+        G_ = self.G.copy()
+        G_.add_edge(u.identifier, v.identifier, weight=on or self._DEFAULT_ARG_WEIGHT)
+
+        if not nx.is_directed_acyclic_graph(G_):
+            raise TypeError('The graph is not DAG with this edge')
+        self.G = G_
+
+        if update_edge:
+            self.edges.append(
+                (u.identifier, v.identifier, on or self._DEFAULT_ARG_WEIGHT)
+            )
+            if isinstance(u, Model) and u not in self.models:
+                self.models.append(u)
+            if v not in self.models:
+                self.models.append(v)
+        return
+
+    def stash_node_output(self, node, output, cache={}):
+        cache[node.identifier] = output
+
+    def get_output(
+        self, node, index: t.Optional[int] = None, one: bool = False, cache: dict = {}
+    ):
+        output = cache[node.identifier]
+        if index is not None:
+            assert isinstance(index, int)
+            if index == -1:
```

Comment:
Can we use another thing to get the whole output, for example, `None` or something else? 
Because this makes me wonder if I want to take the last one. 

WDYT?